### PR TITLE
feat: add resource variables support in builder

### PR DESCRIPTION
**Pull request by TrySound**
Ref https://github.com/webstudio-is/webstudio/issues/2532

Here added Value/Resource tabs to "new/edit variable" panel to support creating both value and http request based variables.

Url, headers and body are expressions with support to all available in scope value and page params variables.

Actual loading of resources in builder will be added later.

Screenshot 2023-12-24 at 13 42 34
Screenshot 2023-12-24 at 13 42 20

## Code Review

- [ ] hi @kof, I need you to do
  - conceptual review (architecture, feature-correctness)
  - detailed review (read every line)
  - test it on preview

## Before requesting a review

- [ ] made a self-review
- [ ] added inline comments where things may be not obvious (the "why", not "what")

## Before merging

- [ ] tested locally and on preview environment (preview dev login: 5de6)
- [ ] updated [test cases](https://github.com/webstudio-is/webstudio/blob/main/apps/builder/docs/test-cases.md) document
- [ ] added tests
- [ ] if any new env variables are added, added them to `.env.example` and the `builder/env-check.js` if mandatory

### CODE REVIEW

**Comment by kof** [comment]
- The "(Expression)" in labels is suboptimal, we probably need some kind of icon or something minimalistic, its fine for the prototype though
- I would guess expression in body allows to access results and do something with them, but what does expression in url does?

**Review by kof — APPROVED** [review]

File: apps/builder/app/builder/features/settings-panel/variables.tsx [hunk 1]
```diff
@@ -108,38 +115,46 @@ const saveVariable = (
...
+        value,
+      };
+      if (typeof value === "string") {
+        variableValue = { type: "string", value };
+      }
+      if (typeof value === "number") {
+        variableValue = { type: "number", value };
+      }
+      if (typeof value === "boolean") {
+        variableValue = { type: "boolean", value };
```
Comment (kof) [inline apps/builder/app/builder/features/settings-panel/variables.tsx:137]: this feels like it shold be generalized in a utility somewhere, I think I saw similar code already before

**Comment by kof** [comment]
@TrySound do we want take expressions in any input in the future? in that case it could be just our standard, then we don't need to explain anywhere one can use expression

**Comment by TrySound** [comment]
> I would guess expression in body allows to access results and do something with them, but what does expression in url does?

We want to parameterize with url params, some constants or access tokens. Those values are passed to generated resources in remix loader too

> do we want take expressions in any input in the future? in that case it could be just our standard, then we don't need to explain anywhere one can use expression

ideally we need a switcher like n8n has instead of popover which cary whole variable management with it. In this specific case we would prefer to have a text field by default and then switch to expression if necessary.

**Comment by kof** [comment]
> We want to parameterize with url params, some constants or access tokens. Those values are passed to generated resources in remix loader too

so url could look like `"/path/" + token` ? 

> ideally we need a switcher like n8n has instead of popover which cary whole variable management with it. In this specific case we would prefer to have a text field by default and then switch to expression if necessary.

seems like we need to think this better through with Mark

**Comment by TrySound** [comment]
> so url could look like "/path/" + token ?


Yes, or "/blog/" + slug

**Pull request merged**